### PR TITLE
Add dropdown for AI teirs

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
## Proposed Changes

* Show a dropdown in the AI lightbox for the user to select what tier they want
* Default to opening lightbox for AI instead of going to checkout
* Have link to support if they want a larger than the max tier

NOTES
* There is an issue with the dropdown when navigating with the keyboard, the link in the last item is not focusable. This is [a deeper issue with the Dropdown component](https://github.com/Automattic/wp-calypso/blob/trunk/packages/components/src/select-dropdown/index.jsx#L358) where only the list item is focusable. The logic seems complex and out of the scope of this PR to fix. I am not entirely sure if this is worth approaching the issue differently (a note below the price rather than a dropdown item) but this is what I have come up with so far
* The highlight animation around the change in request limits in the includes section will be handled separately as this PR was already large enough

## Testing Instructions

1. Checkout this branch or use the Calypso Green live link and go to `/pricing/`
2. Click on "Get" next to the Jetpack AI product and make sure the lightbox opens up with tiers to select
image
3. Open up the dropdown and confirm all the tiers are present
image
4. Switch between them to make sure the price as well as the request limit in the includes section gets updated with each tier
Screenshot 2024-04-30 at 11 41 12 AM
5. Pick a tier and then select "Proceed to checkout" and make sure the correct tier is in checkout
image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/plans/jetpack-plans/product-lightbox/constants.ts [hunk 1]
```diff
@@ -25,24 +28,30 @@ export const PRODUCT_OPTIONS: Record< string, TranslateResult > = {
...
-		comment: 'Get up to 1000 shares',
-	} ),
-	[ PRODUCT_JETPACK_SOCIAL_ADVANCED_BI_YEARLY ]: translate( 'Social Advanced', {
-		comment: 'Get unlimited shares and share as a post by attaching images or videos.',
-	} ),
-	[ PRODUCT_JETPACK_SOCIAL_ADVANCED ]: translate( 'Social Advanced', {
-		comment: 'Get unlimited shares and share as a post by attaching images or videos.',
-	} ),
-	[ PRODUCT_JETPACK_SOCIAL_ADVANCED_MONTHLY ]: translate( 'Social Advanced', {
-		comment: 'Get unlimited shares and share as a post by attaching images or videos.',
```
Comment (CodeyGuyDylan) [inline client/my-sites/plans/jetpack-plans/product-lightbox/constants.ts:44]: These translator comments seemed more like marketing comments so I removed them. Additionally we don't want to translate product names anyway

File: client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx [hunk 2]
```diff
@@ -96,17 +100,18 @@ export const AllItems: React.FC< AllItemsProps > = ( {
 						isSuperseded
 					);
 
-					const isSocialProduct = isJetpackSocialSlug( item.productSlug );
+					const isMultiPlanSelectProduct =
+						isJetpackSocialSlug( item.productSlug ) || isJetpackAISlug( item.productSlug );
```
Comment (CodeyGuyDylan) [inline client/my-sites/plans/jetpack-plans/product-store/items-list/all-items.tsx:104]: The plan is to only add AI for now, stats will be dealt with in a separate PR

**Comment by jjchrisdiehl** [comment]
I'm not sure putting the contact link within the dropdown is the best approach, as we tend to see these variant pickers overused as it is and it can lead to future design issues along with the accessibility issues you noted.

Would it make more sense to add the last link beneath the 'Proceed to Checkout' button? Since (I presume) the intended action is to contact support and abandon the checkout flow anyway, it wouldn't hurt to move the contact button to a more prominent place like:

image

This also has the advantage of making the contact button accessible to keyboards again:
image

**Comment by CodeyGuyDylan** [comment]
> Would it make more sense to add the last link beneath the 'Proceed to Checkout' button?

I had suggested this at first, I'll bring it up again due to the accessibility issues

**Comment by eeeeeevon2023** [comment]
Looks good, some minor tweaks. Let's increase the vertical distance between the "Need more than 1000 requests" section and the CTA. We don't want it to be passed over at a glance as a TOS.

**Comment by CodeyGuyDylan** [comment]
> Looks good, some minor tweaks. Let's increase the vertical distance between the "Need more than 1000 requests" section and the CTA. We don't want it to be passed over at a glance as a TOS.

Thank you for the input @eeeeeevon2023, what do you think of @augurlu's suggestion to not show any link for "more than 1000 requests" until we have a better flow for it? (p1HpG7-rEl-p2#comment-71617)

**Review by elliottprogrammer — APPROVED** [review]
Tested, works great as described. The code looks all good to me!  I don't have any suggestions or corrections.
LGTM!! 👍

File: client/my-sites/plans/jetpack-plans/product-lightbox/constants.ts (hunk 1 above)
Comment (jboland88) [reply client/my-sites/plans/jetpack-plans/product-lightbox/constants.ts:44]: Seems fine to me

File: packages/calypso-products/src/products-list.ts [hunk 3]
```diff
@@ -120,17 +135,19 @@ import type {
...
 
 export const JETPACK_SITE_PRODUCTS_WITH_FEATURES: Record<
-	Exclude< JetpackProductSlug, WPComProductSlug >,
+	Exclude< JetpackProductSlug | JetpackAliasSlug, WPComProductSlug >,
 	Product
 > = {
 	[ PRODUCT_JETPACK_AI_MONTHLY ]: {
 		product_name: PRODUCT_SHORT_NAMES[ PRODUCT_JETPACK_AI_MONTHLY ],
 		product_slug: PRODUCT_JETPACK_AI_MONTHLY,
+		product_alias: PRODUCT_JETPACK_AI_MONTHLY_100,
```
Comment (jboland88) [inline packages/calypso-products/src/products-list.ts:150]: I see you introduced an alias on the monthly plan but not yearly or biennial - is this needed on the other base plan definitions as well?

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx [hunk 4]
```diff
@@ -19,23 +19,24 @@ type PaymentPlanProps = {
...
 
-	const { originalPrice, discountedPrice, discountedPriceDuration, isFetching } = useItemPrice(
-		siteId,
-		product,
-		product?.monthlyProductSlug || ''
-	);
+	const { originalPrice, discountedPrice, discountedPriceDuration, isFetching, priceTierList } =
+		useItemPrice( siteId, product, product?.monthlyProductSlug || '' );
 
+	const currentTier = quantity && priceTierList.find( ( tier ) => tier.maximum_units === quantity );
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx]: The only change I'd suggest here is putting quantity between the min and max
```
max >= qty >= min
```

That would feel a bit more robust as opposed to the strict equality IMO

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx [hunk 5]
```diff
@@ -19,23 +19,24 @@ type PaymentPlanProps = {
...
+	const currentTierPrice = currentTier && currentTier.minimum_price / 100 / 12;
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:48]: I assume the division by 12 is just for annual plans here.
I think that's probably OK for now - as long as we don't shift to showing other plan terms with this code 😅

**Review by jboland88 — COMMENTED** [review]
I left some small feedback on the pricing calculation.

I'm not quite sure I understand the aliasing solution - can you point me to some more context around that solution? I am wondering if there's a way to handle the quantity variance without needing all the additional product definitions for each tier 🤔 

Maybe other solutions would require more dev work?

Not opposed to the aliasing - it just felt a bit confusing to figure it out.

File: packages/calypso-products/src/constants/jetpack.ts [hunk 6]
```diff
@@ -230,6 +245,24 @@ export const JETPACK_AI_PRODUCTS = < const >[
 	PRODUCT_JETPACK_AI_YEARLY,
 ];
 
+export const JETPACK_AI_ALIASES = < const >[
```
Comment (jboland88) [inline packages/calypso-products/src/constants/jetpack.ts:248]: Would be helpful to add a comment here to explain what the aliases are being used for.

**Comment by CodeyGuyDylan** [comment]
> I left some small feedback on the pricing calculation.
> 
> I'm not quite sure I understand the aliasing solution - can you point me to some more context around that solution? I am wondering if there's a way to handle the quantity variance without needing all the additional product definitions for each tier 🤔
> 
> Maybe other solutions would require more dev work?
> 
> Not opposed to the aliasing - it just felt a bit confusing to figure it out.

Pasting from our slack convo:

Just for completion’s sake, I looked at the PR again and remembered the main reason I decided to make constants again :sweat_smile: . It was [this part](https://github.com/Automattic/wp-calypso/pull/89989/files#diff-904f70795869cc713b297c41b0ff369aefdb79be9c29eec01b6a6e4dc5b00c5eR159). I needed a way to add quantity to the selector product itself so I could use it in several places (checkoutUrl, price, etc.) and the use of [slugToSelectorProduct](https://github.com/Automattic/wp-calypso/pull/89989/files#diff-60c4f5ddf04ffcd11db73b26b177a483a573c08d466e9125f4083e11eeb0af44R68) matches with that list of constants to create a base product the [itemToSelectorProduct](https://github.com/Automattic/wp-calypso/pull/89989/files#diff-2ad10827720e63b2d6aefe8cb912733aab4fc45e2dc3f0ed3902d780b9c8ff40L130) function can use to add more properties.
Without having the aliases in the JETPACK_SITE_PRODUCTS_WITH_FEATURES array, there was no way to get quantity into the selector product other than adding a second argument to slugToSelectorProduct. Which isn’t the end of the world, but that function is named to indicate you can get the product with the slug only, adding more arguments to that seems to defeat the purpose of having it in the first place
Anyway I hope that argument makes sense, I’ll address your comments on the PR and we can get it shipped today hopefully :slightly_smiling_face:

FYI the other option I initially had was to send the slug with the quantity to slugToSelectorProduct and split it up before calling itemToSelectorProduct and adding the quantity manually. Seemed hacky to me and I didn’t like it.. However that would be less code so I’m willing to revisit that if that sounds like it would be a better option

File: packages/calypso-products/src/products-list.ts (hunk 3 above)
Comment (CodeyGuyDylan) [reply packages/calypso-products/src/products-list.ts:150]: Nice catch, the yearly and bi-yearly ones should have them too. It's not entirely necessary for what we are using them for right now but it should be added for consistency

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx (hunk 4 above)
Comment (CodeyGuyDylan) [reply client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx]: 🤔 

I am not sure if that is necessary here. The quantity comes from a tier list that is generated from the `priceTierList` which has set values we use. I don't see any way that a value not exactly the same as an item in the list gets passed to this component. 

However, it's possible some code later on gets added where that is a possibility and uses the same component (though not likely, as this component is specific to the lightboxes). I can get that added since it's not an expensive computation and for completeness 😄

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx (hunk 5 above)
Comment (CodeyGuyDylan) [reply client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:48]: It's probably find since we show annual prices everywhere. I don't see that changing anytime soon 😄

File: packages/calypso-products/src/constants/jetpack.ts (hunk 6 above)
Comment (CodeyGuyDylan) [reply packages/calypso-products/src/constants/jetpack.ts:248]: Will do!

**Comment by CodeyGuyDylan** [comment]
Will need a partial re-review on this. Mostly just functionality (works fine for me) because I refactored the lightbox to extract the option components to their own files to make the main file more readable

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx [hunk 7]
```diff
@@ -19,23 +19,33 @@ type PaymentPlanProps = {
...
-		product?.monthlyProductSlug || ''
-	);
+	const { originalPrice, discountedPrice, discountedPriceDuration, isFetching, priceTierList } =
+		useItemPrice( siteId, product, product?.monthlyProductSlug || '' );
+
+	const currentTier =
+		quantity &&
+		priceTierList.find( ( tier ) => {
+			if ( tier.maximum_units ) {
+				return quantity >= tier.minimum_units && quantity <= tier.maximum_units;
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:42]: Ah, I did not consider - if we check against both min an max units here - do we also need to verify that `tier.minimum_units` exists too?

File: client/my-sites/plans/jetpack-plans/product-lightbox/product-select.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+interface ProductSelectProps {
+	product: SelectorProduct;
+	siteId: number | null;
+	onChangeProduct: ( product: SelectorProduct | null ) => void;
+}
+
+const ProductSelect: FC< ProductSelectProps > = ( { product, siteId, onChangeProduct } ) => {
+	const dispatch = useDispatch();
+	const isMultiOptionProduct = isJetpackSocialSlug( product.productSlug );
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/product-lightbox/product-select.tsx:18]: Is AI needed here too?

Comment (jboland88) [reply client/my-sites/plans/jetpack-plans/product-lightbox/product-select.tsx:18]: Oh, wait - I see now that this is the multiple choice selection :)

File: client/my-sites/plans/jetpack-plans/product-lightbox/quantity-dropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	onChangeProduct: ( product: SelectorProduct | null ) => void;
+}
+
+const QuantityDropdown: FC< QuantityDropdownProps > = ( { product, siteId, onChangeProduct } ) => {
+	const dispatch = useDispatch();
+	const listPrices = useItemPrice( siteId, product, product?.monthlyProductSlug || '' );
+	const isProductWithTierList = isJetpackAISlug( product.productSlug );
+
+	const tierOptions = useMemo( () => {
+		if ( ! isJetpackAISlug( product.productSlug ) ) {
```
Comment (jboland88) [inline client/my-sites/plans/jetpack-plans/product-lightbox/quantity-dropdown.tsx]: Nitpick: could use `isProductWithTierList` here

**Review by jboland88 — APPROVED** [review]
Tested, works functionally for me.
Code LGTM.

File: client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx (hunk 7 above)
Comment (CodeyGuyDylan) [reply client/my-sites/plans/jetpack-plans/product-lightbox/payment-plan.tsx:42]: In the types, minimum_units is labelled as a required value. Because of that, I don't think we need to?

File: client/my-sites/plans/jetpack-plans/product-lightbox/quantity-dropdown.tsx (hunk 9 above)
Comment (CodeyGuyDylan) [reply client/my-sites/plans/jetpack-plans/product-lightbox/quantity-dropdown.tsx]: Will get this added and ship 😄

Comment (CodeyGuyDylan) [reply client/my-sites/plans/jetpack-plans/product-lightbox/quantity-dropdown.tsx]: Nice catch
